### ISSUE TITLE
Evaluate for Runner

### ISSUE THREAD

**Issue by Scitator** [catalyst-team/catalyst#1186]
## 🚀 Feature Request

The `evaluate_loader` method for Python API. Similar to `.train` and `.predict_loader`

### Proposal

Possible use case
```python
import os
from torch import nn, optim
from torch.utils.data import DataLoader
from catalyst import dl, utils
from catalyst.data.transforms import ToTensor
from catalyst.contrib.datasets import MNIST

model = nn.Sequential(nn.Flatten(), nn.Linear(28 * 28, 10))
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.02)

loaders = {
    "train": DataLoader(
        MNIST(os.getcwd(), train=True, download=True, transform=ToTensor()), batch_size=32
    ),
    "valid": DataLoader(
        MNIST(os.getcwd(), train=False, download=True, transform=ToTensor()), batch_size=32
    ),
}

runner = dl.SupervisedRunner(
    input_key="features", output_key="logits", target_key="targets", loss_key="loss"
)
# model training
runner.train(
    model=model,
    criterion=criterion,
    optimizer=optimizer,
    loaders=loaders,
    num_epochs=1,
    callbacks=[
        dl.AccuracyCallback(input_key="logits", target_key="targets", topk_args=(1, 3, 5)),
        dl.PrecisionRecallF1SupportCallback(
            input_key="logits", target_key="targets", num_classes=10
        ),
        dl.AUCCallback(input_key="logits", target_key="targets"),
        # catalyst[ml] required ``pip install catalyst[ml]``
        # dl.ConfusionMatrixCallback(input_key="logits", target_key="targets", num_classes=10),
    ],
    logdir="./logs",
    valid_loader="valid",
    valid_metric="loss",
    minimize_valid_metric=True,
    verbose=True,
    load_best_on_end=True,
)

loader_metrics = runner.evaluate_loader(
    loader=loaders["valid"]), 
    callbacks=[
        dl.AccuracyCallback(input_key="logits", target_key="targets", topk_args=(1, 3, 5)),
        dl.PrecisionRecallF1SupportCallback(
            input_key="logits", target_key="targets", num_classes=10
        ),
    ])
```

### Alternatives

The whole method could be easily done with the `.train` approach, but for a more user-friendly API – why should not we add a simplified alias?

### Checklist
- [x] feature proposal description
- [x] motivation
- [x] extra proposal context / proposal alternatives review
  

### FAQ
Please review the FAQ before submitting an issue:
- [x] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [x] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [x] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [x] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [x] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

**Comment by AlekseySh** [comment]
In general, I like this proposal, but I have one point to add:
What if we have trained our model via `train.py`, but then we are going to evaluate this model in `valid.py`?.
In this case, `Runner` objects would be different. But the model was defined only in the first runner.
And I guess, that we also don't want to bring all the trained parameters to evaluation.
So, we should separate it somehow.


Possible proposal (make `evaluate_loader` as a separate function with `model` parameter): 

```
loader_metrics = evaluate_loader(
    model = model,
    loader=loaders["valid"]), 
    callbacks=[
        dl.AccuracyCallback(input_key="logits", target_key="targets", topk_args=(1, 3, 5)),
        dl.PrecisionRecallF1SupportCallback(
            input_key="logits", target_key="targets", num_classes=10
        ),
    ])

```
It is also can be an additional method of `Runner`, but it should work without thain-related parameters. 


Or you can suggest something else, but I am sure, that we have to separate train and val params into different entities.

**Comment by leoromanovich** [comment]
@AlekseySh Hi!
Do you mean functional like that, but inside Supervised runner?

```python
from typing import Any, Callable, Dict, List, Mapping, OrderedDict
import catalyst
from catalyst import dl
from catalyst.typing import Criterion, Model
import torch
import torch.nn as nn
from torch.utils.data import DataLoader, Dataset


class CustomRunner(dl.SupervisedRunner):
    def evaluate_loader(self,
                        model: nn.Module,
                        loader: DataLoader,
                        criterion: Criterion,
                        callbacks: List,
                        ):
        self.train(
            model=model,
            loaders=OrderedDict([('valid', loader)]),
            num_epochs=1,
            verbose=True,
            callbacks=callbacks,
            valid_loader='valid',
            seed=42,

        )

        return self.loader_metrics


class ToyModel(nn.Module):
    def __init__(self, in_features, out_features):
        super().__init__()
        self.model = nn.Sequential(
            nn.Linear(in_features, out_features)
        )

    def forward(self, x):
        return self.model(x)


class ToyDataset(Dataset):
    def __init__(self):
        self.data = torch.randn(size=(10, 5))
        self.targets = torch.randint(0, 1, size=(10,))

    def __len__(self):
        return len(self.targets)

    def __getitem__(self, index):
        return {'features': self.data[index],
                'targets': self.targets[index]}


loader = DataLoader(ToyDataset(), batch_size=4)
model = ToyModel(5, 1)
runner = CustomRunner()
loader_metrics = runner.evaluate_loader(
    model=model,
    loader=loader,
    callbacks=[
        dl.AccuracyCallback(target_key="targets",
                            input_key="logits",
                            topk_args=[1]),
    ]
)
print(loader_metrics)

```

**Comment by Scitator** [comment]
🤔 @leoromanovich , I like your implementation (with, maybe, a few extra parameters for seed/logdir/etc and a few docs).
Could you please contribute it? From the Runner API perspective, it looks good for me.

**Comment by leoromanovich** [comment]
@Scitator Sure! 
AlekseySh and I are discussing the right way of implementation.

**Comment by Scitator** [comment]
which one? 😂

**Comment by leoromanovich** [comment]
Maybe more common, on the higher level of inheritance of Runners :)

**Comment by Scitator** [comment]
I am now sure, that it would be correct...

**Comment by Scitator** [comment]
I mean, you could make an extra functional wrapper for model-loader evaluation, but the Runner API should looks like the top one, I think.

**Comment by leoromanovich** [comment]
I see. It seems like `evaluate_loader` should be a simple and exclusive method of `SeupervisedRunner`. Am I right?

**Comment by Scitator** [comment]
From the Runner API perspective - it should be the `Runner` method.
I mean, only the Runner knows how we would like to feed the data into the model and how we would like to compute the metrics during custom `bathc_handle`

**Comment by leoromanovich** [comment]
Got it. I'll prepare a method for Runner. =)

**Comment by AlekseySh** [comment]
I mean if we add the eval method to Runner, then it will no affect to HydraRunner and ConfigRunner. Is this expected behavior? @Scitator

**Comment by Scitator** [comment]
Yup, that's the Python API, while Config/Hydra - are other ones. 
Nevertheless, we could try to inherin Config/Hydra from the Runner one... or just make a `.train`, `.evaluate`, `.predict` as extra `IPythonRunner` (we should choose better name) and make the following changes:
-  `Runner(IPythonRunner) & SupervisedRunner(IPythonRunner, ISupervisedRunner, IRunner)`
- `ConfigRunner(IPythonRunner, IRunner) & SupervisedConfigRunner(IPythonRunner, ISupervisedRunner, IRunner)`

What do you think about it?

**Comment by leoromanovich** [comment]
If it will be smthg like this, @Scitator what do you think?

```python 
class Runner(IRunner):
    ...
    def evaluate_loader(
        self,
        model: Model,
        loader: DataLoader,
        callbacks: "Union[List[Callback], OrderedDict[str, Callback]]",
        seed: int = 42,
        logdir: str = None,
        verbose: bool = False,
        **kwargs
    ) -> Dict:
        """
            Evaluates data from loader with given model and returns obtained metrics.
            Args:
                loader: loader to predict
                model: model to use for prediction
                callbacks: list or dictionary with catalyst callbacks
                seed: random seed to use before prediction
                logdir: path to output directory
                verbose: if `True`, it displays the status of the evaluation to the console.


            Returns:
                Dict with metrics counted on the loader.
        """
        self.train(
            model=model,
            loaders=OrderedDict([("valid", loader)]),
            num_epochs=1,
            verbose=verbose,
            callbacks=callbacks,
            valid_loader="valid",
            seed=seed,
            logdir=logdir,
            **kwargs
        )

        return self.loader_metrics

```

**Comment by Scitator** [comment]
It looks good from my point of view 👍 
The only thing, that would be definitely required: different test cases. For example, I am not sure we need to use `CheckpointCallback` (which is usually turned on if `logdir` is specified) to save our model. Or we should write proper docs to "Do not mess train and evaluate logdirs :) ", something like that.

**Comment by Scitator** [comment]
btw, it would be also great if we could update our main `Readme.md` example with `runner.evaluate_loader` :)
Looking forward to see your contribution ;)

**Comment by Scitator** [comment]
hi there,
@leoromanovich do you need any help with the implementation? is there any progress for opening a PR? ;)
